### ISSUE TITLE
Request: Collapse functions

### ISSUE THREAD

**Issue by ankith26** [thonny/thonny#1090]
I am using thonny, I really like it now.
This is just an idea, how about introducing function collapse feature where you can minimise the functions you do not want to see.

**Comment by aivarannamaa** [comment]
Thanks for the idea!

Currently I don't consider this a priority feature for Thonny's target group, but I'll keep it in the TODO list.

**Comment by mrssmcmaster** [comment]
On the contrary, I (as a teacher) would find this very useful for explaining to my students that once a function has been written, it is no longer necessary to know what it does. Being able to collapse it helps with that.

**Comment by aivarannamaa** [comment]
@mrssmcmaster, good point! I never thought of it.

**Comment by mjarboe5** [comment]
> Thanks for the idea!
> 
> Currently I don't consider this a priority feature for Thonny's target group, but I'll keep it in the TODO list.

Would be GREAT to have!!

**Comment by ryanomatic** [comment]
+1 This would be fantastic!

**Comment by rustyjuniper** [comment]
+1 I really would like to get the code folding feature. I just discovered, geany editor can help out here

**Comment by palieux** [comment]
This would be great to have. My son uses Thonny and he loves it. However when learning nested if functions, we had to switch to a different IDE where we could collapse the underlying code. This helped him understand scope and nesting better.
Keep up the great work!

**Comment by Splashaudio** [comment]
Would love to be able to do this. Will it happen?

**Comment by aivarannamaa** [comment]
@palieux, have you tried explaining nested if-blocks using the "nicer" debugger (Ctrl+F5)? I believe this can explain the meaning of if-blocks even better than collapsing the bodies.

**Comment by DIT-Jack** [comment]
ohh get on with it, add code folding

**Comment by pak-man** [comment]
C´mon, we need this. After two years this very basic and quite essential feature is still missing.

**Comment by aivarannamaa** [comment]
@DIT-Jack and @pak-man, c'mon, you want a feature, why don't you make a pull request? Or do you feel you've received less that you've paid for?

**Comment by kingandaval** [comment]
> @DIT-Jack and @pak-man, c'mon, you want a feature, why don't you make a pull request? Or do you feel you've received less than you've paid for?

@aivarannamaa I am only just learning python now and using Thonny as my IDE but collapsing the code would be useful for people new to python. Sadly I cannot contribute to the Thonny project but maybe one day I can submit a pull request ;-)

**Comment by The-Salty-Medic** [comment]
Happy December.  I would also love to see this feature added as a way to tidy up the code while I'm working on things.  I agree with a prior comment about folding up functions / lists / conditional blocks.

**Comment by B0-B** [comment]
@aivarannamaa that's a weird argument, you are offering an open-source software as an organization and aren't able to implement a simple but very important feature and your argument is to fork it? Why should someone use your branch then anyway?

**Comment by MatrixMike** [comment]
So code folding is not here, yet

**Comment by coffandro** [comment]
> @aivarannamaa that's a weird argument, you are offering an open-source software as an organization and aren't able to implement a simple but very important feature and your argument is to fork it? Why should someone use your branch then anyway?

Just curious but how do you know if this feature would be "simple", it entirely depends on how just about everything else works and is implemented. there's a low chance it'd even require a full rewrite thought most likely not but at the same time it isn't really an *essential* feature now is it?

**Comment by AL-M0ND** [comment]
+1 would definitely be a great feature, would save me time scrolling around my long docstrings 😂

**Comment by csunwc** [comment]
+1 I would find the code folding a super helpful feature for reading codes.

**Comment by pak-man** [comment]
I gave up any hope and I am gladly using vscode + rshell instead. Works wonder, and I feel happy I abandoned that clunky piece of ... of editor.

**Comment by csunwc** [comment]
@pak-man I feel your frustrations. How do you use rshell to upload codes?  Do you use the instructions here...   https://github.com/dhylands/rshell

**Comment by pak-man** [comment]
Yes i used that documentation to make some custom commands for VSCode. I like rshell. It´s very robust and features like sync are very very neat. You can also script in it Python (import rshell ...) which is very convenient.

Another useful bit is [pye](https://github.com/robert-hh/Micropython-Editor/tree/master): it´s a text editor that runs on the micropython board itself. Since I learned it I could not do without. It´s just damn convenient to be able to quickly change a local file without having to upload/download. Just rsync back your changes with rshell when you´re done editing locally.

upysh is also pretty handy for dealing with local files, and ugit also when the board has IP.
